### PR TITLE
GHC 9.8 compat and CI

### PR DESCRIPTION
**Pull request by ulysses4ever**
**Template Β: This PR does not modify `cabal` behaviour (documentation, tests, refactoring, etc.)**

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by ulysses4ever** [comment]
I've done basic dumb work to update the code and CI for 9.8 but there are some actual failures that need to be looked into. Help appreciated.

**Comment by ulysses4ever** [comment]
I fixed the actual regressions (due to -Wx-partial) locally and now I'm wrangling with allow-newer.

**Comment by ulysses4ever** [comment]
Timed out while fixing all those heads and tails. Converting to draft for now.

**Comment by ulysses4ever** [comment]
All right, I issue an official call for help: both in Matrix and here. The failing test is here: https://github.com/haskell/cabal/actions/runs/6536363720/job/17747928752?pr=9330

It's about a hash mismatch in `Cabal-tests/tests/UnitTests/Distribution/Utils/Structured.hs`. Anyone has experience with this? Currently, there's one CPP clause with min base already. Should I add another one?
https://github.com/haskell/cabal/blob/9f373253b9032ae32ead9f5ca17e49d192f9b8b9/Cabal-tests/tests/UnitTests/Distribution/Utils/Structured.hs#L28-L33

I had a cursory look over what this `Structured` business does and, to be honest, I'm not positive, but it looks like it tries to control ABI changes in how datatypes look like to avoid troubles with serialization when this changes...

**Comment by ulysses4ever** [comment]
Gershom greenlights adding some CPP to handle this apparent change in how GHC 9.8 codegens data types.

**Comment by ulysses4ever** [comment]
More of the same `structureHash` unhappiness...
https://github.com/haskell/cabal/actions/runs/6538444785/job/17754390522?pr=9330

These two guys:
https://github.com/haskell/cabal/blob/eece442d47b7e87aa2afb8f3bdc8c35b77901d67/cabal-install/tests/UnitTests/Distribution/Client/FileMonitor.hs#L34C1-L35

**Comment by ulysses4ever** [comment]
I'm a bit lost with the current failing test tbh: https://github.com/haskell/cabal/actions/runs/6563593296/job/17828182063?pr=9330

**Comment by Mikolaj** [comment]
This is just a different number of dots after the output lines. Do they come from cabal or GHC?

**Comment by ulysses4ever** [comment]
O, thanks for the pointer, Mikolaj: I completely missed this change. I think this is due to #8929. It looks like it manifests only with GHC 9.8. Presumably, because 9.8 is first to ship with Cabal containing this patch (3.10.2)? I'm not sure why we use Cabal shipped with GHC in the test suite, but I presume because of how `custom-setup` works? The failing test (`AutogenModulesToggling`) does contain a `custom-setup` component.

**Comment by ulysses4ever** [comment]
What would be the way forward? I'm thinking:

1. Disable all current `custom-setup` tests for GHC >= 9.8.
2. Create a copy of those tests and disable them for GHC < 9.8 and update the output.

Quick grepping suggests that there's about 45 tests with `custom-setup`.

Another idea would be to add upper bounds for `Cabal` in those tests. I'm not sure this will work, in particular, whether the tests environment allows to build a `Cabal` not available in the boot packages. But I could investigate...

**Comment by Mikolaj** [comment]
All avenues you listed make sense.

> Disable all current custom-setup tests for GHC >= 9.8.

Brutal, but effective. Who knows, we may want to change the tests further in the future (this copy or yet another one), porting them to the new Hooks build mode if it gets accepted into cabal.

**Comment by mpickering** [comment]
> What would be the way forward? I'm thinking:
> 
>     1. Disable all current `custom-setup` tests for GHC >= 9.8.
> 
>     2. Create a copy of those tests and disable them for GHC < 9.8 and update the output.
> 
> 
> Quick grepping suggests that there's about 45 tests with `custom-setup`.
> 
> Another idea would be to add upper bounds for `Cabal` in those tests. I'm not sure this will work, in particular, whether the tests environment allows to build a `Cabal` not available in the boot packages. But I could investigate...

I think the right thing to do is to refine the test so that not the whole stdout is compared. What is the test supposed to test (and what is the purpose of comparing the stdout).

**Comment by ulysses4ever** [comment]
So, the failing test, `AutogenModulesToggling`, was added as a part of #8690 with

```
skipUnlessGhcVersion ">= 9.7"
```

so it had never fired before this PR, I think. Then, it was modified to use three dots by https://github.com/haskell/cabal/pull/8929 --- by the way of `grep`, I think. If this is the only test that has this kind of trouble, then I just modify the reference output, I think. It should Just Work™, since it doesn't get called with GHCs <9.8 anyway.

**Comment by ulysses4ever** [comment]
Quite unfortunately, there is one line that differs not only in the number of dots but in an actual behavior:

https://github.com/haskell/cabal/actions/runs/6563593296/job/17828182063?pr=9330#step:18:463

At this point I need to summon the author of the test, @sergv, to ask: are you concerned that with GHC 9.8 your patch (#8690) have the opposite effect to what was recorded initially:

```
-The module says: Prebuilt module, don't use in production
+The module says: Real module, ship to production
```
And also, why did  you use `skipUnlessGhcVersion ">= 9.7"`? Did you test it with development version of GHC or something?

**Comment by sergv** [comment]
I expected #8690 to be included in Cabal package within GHC 9.8.1. Without patch available in builtin Cabal package the test doesn't make sense, hence the bounds check.

Current output is puzzling - when -fgenerate is passed a new module is generated, good, but when asked to not generate files the predefined file should have been picked up.

> Did you test it with development version of GHC or something?
As mentioned in #8690 I tested with patched ghc, however the reason for bound is explained previously - test needs patched Cabal bundled with ghc.

**Comment by ulysses4ever** [comment]
I'm trying to expect-broken this test in the meantime but I'm failing to do so.

Attempt one:
```haskell
main = cabalTest . recordMode RecordMarked . expectBroken 9330 $ do                                                              
│ skipUnlessGhcVersion ">= 9.7"                                                                                                  
│ cabal "v2-run" ["-fgenerate", "autogen-toggle-test"]                                                                           
│ cabal "v2-run" ["-f-generate", "autogen-toggle-test"]  
```
Attempt two:
```haskell
main :: IO ()                                                                                                                                           
main = cabalTest . recordMode RecordMarked . expectBroken 9330 $ do                                                                                     
│ skipUnlessGhcVersion ">= 9.7"                                                                                                                         
│ cabal "v2-run" ["-fgenerate", "autogen-toggle-test"]                                                                                                  
│ r <- cabal' "v2-run" ["-f-generate", "autogen-toggle-test"]                                                                                           
│ assertOutputContains "Prebuilt module, don't use in production" r     
```
Both fail with kilometers of output (~400 lines each). Would appreciate any pointers.

**Comment by ulysses4ever** [comment]
@sergv thank you for the explanation. If you have some more time for this: could you confirm that GHC 9.8.1 behaves as you expected? The patch was merged in February, and I'd certainly expect it to ship in Cabal 3.10.2, which was shipped with GHC 9.8.1.

**Comment by sergv** [comment]
@ulysses4ever I've taken a look at 9.8 behaviour and cannot make sense of it. At least I managed to locally reproduce the same results as in https://github.com/haskell/cabal/actions/runs/6563593296/job/17828182063?pr=9330#step:18:456.

Notice, however, that there's not just difference in `The module says: Real module, ship to production` line. There are differences in the number of dots, the ellipsis, which were fixed to 3 after my #8690. At least superficially it looks like GHC does not include fix for the number of ellipsis nor #8690.

Having checked out 15a0010461c3d181f30fbbf5980bf75945d6df2a, the `Cabal-syntax-v3.10.2.0` tag, I don't see the changes made by #8690 therefore the change didn't make it into 9.8.1. So I suggest to bump test's bound on GHC to be `>= 9.9` (not sure whether current dev version has recent enough Cabal though).

**Comment by sergv** [comment]
I've made a PR to disable `AutogenModulesToggling` test on 9.8, hopefully it will work with 9.10 in the future. If you can, please review https://github.com/haskell/cabal/pull/9398.

**Comment by ulysses4ever** [comment]
@sergv thanks a lot for your analysis, is very helpful! Our backporting practices are not good enough, and that kind of thing happens once in a while. Usually every new major version of GHC would include a new major version of Cabal, so this was less visible until GHC 9.8, which parted with this rule and included a minor relrase of Cabal (3.10.2).

I agree that it's best to postpone this test to the next GHC release.

As for your new PR, it would be much easier for me to include this change in this current PR than to see your PR through and rebase (finding reviewers is not always easy, and after that there's the two-day delay, and rebasing....). But if you insist on a separate PR, we could go that way.

**Comment by sergv** [comment]
I don't insist on separate PR, please feel free to work in your own PR.

**Comment by ulysses4ever** [comment]
Oh no, the double-directory back strikes again, only now it's a double time! https://github.com/haskell/cabal/actions/runs/6845077415/job/18609772006?pr=9330#step:15:271

**Comment by Mikolaj** [comment]
Yay! Cumulation. Let me wipe out the cache in that case...

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by Mikolaj** [comment]
Wow, and this is one of those times when cache reset does not help (or the dice rolled wrong again).

**Review by Mikolaj — APPROVED** [review]
Yipee!

**Comment by ulysses4ever** [comment]
It was a long ride… I'll double check that commit history makes sense (I edited it several times) and then apply the merge label.

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -62,6 +62,12 @@ jobs:
 
       - uses: actions/checkout@v4
 
+      - uses: haskell-actions/setup@v2
+        id: setup-haskell
+        with:
+          ghc-version: ${{ matrix.ghc }}
+          cabal-version: '3.10.1.0'
```
Comment (ulysses4ever) [inline .github/workflows/validate.yml:69]: Should be latest (3.10.2.1?)

File: Cabal/src/Distribution/Simple/GHC.hs [hunk 2]
```diff
@@ -171,14 +171,14 @@ configure verbosity hcPath hcPkgPath conf0 = do
       (userMaybeSpecifyPath "ghc" hcPath conf0)
   let implInfo = ghcVersionImplInfo ghcVersion
 
-  -- Cabal currently supports ghc >= 7.0.1 && < 9.8
+  -- Cabal currently supports ghc >= 7.0.1 && < 9.10
```
Comment (ulysses4ever) [inline Cabal/src/Distribution/Simple/GHC.hs:174]: Commit message doesn’t make sense

File: .github/workflows/validate.yml (hunk 1 above)
Comment (Mikolaj) [reply .github/workflows/validate.yml:69]: Probably doesn't matter much, but yes, I guess?

Comment (ulysses4ever) [reply .github/workflows/validate.yml]: I put back `latest` as it was at some point.

Comment (Mikolaj) [reply .github/workflows/validate.yml]: Next time we change it from `latest`, let's write in a comment (not, e.g., commit or PR description) why we do. :)

**Comment by ulysses4ever** [comment]
Took only two months… 🥲

**Pull request merged**